### PR TITLE
feature(hooks): Adds elgg_filter as shorthand for triggering a hook

### LINKED ISSUE #8433: Filter as sugar for passing a value through a plugin hook
Hooks are great for making code collaborative but the syntax is ugh:
- `elgg.trigger_plugin_hook` and `elgg_trigger_plugin_hook` are both long
- requires 2 names for the hook/type and no one can agree on how to use these
- Params is the 3rd arg even though we often don't need it

I suggest adding a "filter" function with a more pleasant API:

``` php
$foo = elgg_filter('myPlugin:foo', $foo);
```

is the equivalent of:

``` php
$foo = elgg_trigger_plugin_hook('filter', 'myPlugin:foo', null, $foo);
```

This would reduce boilerplate and clarify meaning to encourage devs to make more values alterable.

This would also ease the pain of making module methods swappable:

``` js
require('elgg/booted'); // ensure hooks registered #8365
return {
  foo: elgg.filter('myModule:foo', function() {
    // default implementation
  })
};
```

### PR DESCRIPTION
**Pull request by mrclay**
elgg_filter() and elgg.filter() are simple shortcuts for passing a value through the ‘filter’ plugin hook.

Fixes #8433

### CODE REVIEW

File: docs/guides/javascript.rst [hunk 1]
```diff
@@ -302,7 +302,22 @@ Emit a hook event in the event system.
...
+
+Pass a value through the "filter" hook.
+
+.. code:: js
+
+    // allow other plugins to alter value
+    value = elgg.filter('my_plugin:value', value);
+
+    // define a class method that plugins can alter
+    Foo.prototype.init = elgg.filter('Foo::init', function () {
```
Comment (ewinslow) [inline docs/guides/javascript.rst]: Yikes :/ This feels a little too fancy. In particular plugin hook configuration is local/runtime info while I usually associate class definitions with global/static/compile time. I realize these can be conflated/mixed/matched in JS, but the dependency graph here feels backwards.

**Comment by ewinslow** [comment]
So this is cool, and definitely easier, but did you have a particular thing you wanted to use it for?

File: docs/guides/javascript.rst (hunk 1 above)
Comment (mrclay) [inline docs/guides/javascript.rst]: Yeah, don't need this.

**Comment by mrclay** [comment]
Have a better idea...

**Comment by mrclay** [comment]
No particular need at the moment, it just removes several small roadblocks to filtering values:
* What do I name the hook?
* What do I name the type?
* Which argument does my value go in?
* If I don't have params, what should I give?

Shorter term I'm going to suggest we add a convention to the docs.

**Pull request closed without merging**